### ISSUE TITLE
Pass passwords to alertmanager when using kube-prometheus

### ISSUE THREAD

**Issue by soosap** [coreos/prometheus-operator#1783]
Hi I wonder how I can pass the **smtp_auth_password** into the `alertmanager.yaml` config file when using `kube-prometheus`. I came to the conclusion that it is only possible to store the password in clear text. That is a problem for me as I am committing the repo to version control.

Any solution, any idea to circumvent this shortcoming?

**Comment by brancz** [comment]
There are two things you would want to keep in mind when doing this:

* you should not check in the generated manifests with jsonnet, as that would mean that you are checking in the same password, just base64 encoded.
* assuming you have some way to access secrets somehow via the command line, then you can use jsonnet's external variables (aka `extVar`), to pass the secret to the jsonnet code: https://jsonnet.org/learning/tutorial.html#Parameterize-Entire-Config

That way you will never have to check in secret material.

**Comment by soosap** [comment]
Hi @brancz, I had a look at this and it seems like a interesting solution. Unfortunately, it does not fit well into the helm ecosystem that we use to deploy kubernetes workloads. Your suggestion is to store `.jsonnet` files the source code and have your configuration management tool pull the secrets and render the `.jsonnet` templates in an adhoc style.

Then what do I do with the rendered config file (which includes secrets and cannot be checked into version control). Not sure what to do FROM the point where I have the rendered config TO the point of deploying the workloads in Kubernetes w/ helm. Maybe you can clarify whether it is even possible to use jsonnet along w/ helm.

I have deployed many such tools like **Grafana** or **Portus** and they allow setting configuration through environment variables. i.e. in grafana the environment variable `GF_SECURITY_ADMIN_PASSWORD` overwrites the corresponding piece of config in the config file (support was added just recently). The same methodology is used by **Portus** to provide secret information. Trying to understand why alertmanager forces me to adapt a entirely new system to provide secrets into configuration. While I understand the benefits and the power of jsonnet, I am concerned about the lack of integration w/ other long established tools such as helm.

**Comment by brancz** [comment]
While it's possible to integrate with helm, it's unnecessary to do so. What is generated via jsonnet can be applied against a cluster with a simple `kubectl apply -f`. No helm needed, but you can also use the result and package it up with helm, if that's what you want to do.

May I ask how you handle secrets with helm today?

The discussion of using environment variables for configuration is not really one to have in the Prometheus Operator, but rather in the Alertmanager itself, and aside from that using environment variables for secrets can be harmful (see [here](http://movingfast.io/articles/environment-variables-considered-harmful/) and [here](https://diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/)).

**Comment by soosap** [comment]
Ok I was able to give this some thought. I could deploy workloads in helm using:
```sh
$ jsonnet values.jsonnet > /tmp/values.json
$ helm install --name=prometheus --namespace=default --values=/tmp/values.json ./prometheus
$ rm /tmp/values.json
```

Have not quite understood how to use jsonnet extVar, but will explore this to decide whether it could work with the existing infrastructure that we have.

Ahh ja good question, I use Vault together w/ policies that are assigned to the default service account in a namespace. That means containers have access to a subset of vault that they are allowed to see. In my setup I do not pass the actual password as environment variable. I pass the vault path as the environment variable. Inside the container's entrypoint script these paths are replaced by the actual secrets prior starting pid1. The script makes requests against the vault server for the paths it has received through environment variables.

example:
```Dockerfile
ENV GF_SECURITY_ADMIN_PASSWORD secret/default/prometheus/GF_SECURITY_ADMIN_PASSWORD
```

Not a security expert really to understand if this approach is also harmful. But many of the arguments in the article you posted do not apply in the approach we took. However, it comes at the cost that I need to tweak the base images that I use by adding a vault entrypoint script to the images. That's a couple of mins of work since I know exactly how to tweak the images. Unfortunately, alertmanager being so opinionated forces me to explore a different solution although the setup that we have worked quite well for us thus far.
